gcc/rust/ast/rust-ast.cc

```diff
@@ -3310,21 +3310,40 @@ Module::process_file_path ()
 
   // This corresponds to the path of the file 'including' the module. So the
   // file that contains the 'mod <file>;' directive
-  std::string including_fname (outer_filename);
+  std::string including_fpath (outer_filename);
 
   std::string expected_file_path = module_name + ".rs";
   std::string expected_dir_path = "mod.rs";
 
-  auto dir_slash_pos = including_fname.rfind (file_separator);
+  auto dir_slash_pos = including_fpath.rfind (file_separator);
   std::string current_directory_name;
+  std::string including_fname;
 
-  // If we haven't found a file_separator, then we have to look for files in the
-  // current directory ('.')
+  // If we haven't found a file_separator, then we may have to look for files in
+  // the current directory ('.')
   if (dir_slash_pos == std::string::npos)
-    current_directory_name = std::string (".") + file_separator;
+    {
+      including_fname = std::move (including_fpath);
+      including_fpath = std::string (".") + file_separator + including_fname;
+      dir_slash_pos = 1;
+    }
+  else
+    {
+      including_fname = including_fpath.substr (dir_slash_pos + 1);
+    }
+
+  auto path_string = filename_from_path_attribute (get_outer_attrs ());
+
+  if (!path_string.empty () || !module_scope.empty ()
+      || including_fname == "mod.rs" || including_fname == "lib.rs"
+      || including_fname == "main.rs" || including_fname.size () < 3
+      || including_fname.compare (including_fname.size () - 3, 3, ".rs"))
```

Comment:
this condition is very big - why is the size of `including_fname` a determining factor here? also, can this condition be moved in an outer static function?